### PR TITLE
Add One Time Passcode Input

### LINKED ISSUE #255: Create OTP component(s)
With a headless component library like: https://github.com/guilhermerodz/input-otp

### PR DESCRIPTION
**Pull request by ThomasSeaver**
## Description

Creates a component for one time passcode input, based on a [primitive](https://input-otp.rodz.dev/). Mostly just took the examples and modified it to use our colors / patterns, used SVGs instead of divs, plus a little class modification since the outlines got weird. I skipped adding the carat because it felt awkward and I don't think it shows up in the current example, but I can add that if we feel like it's necessary beyond the highlighting.

Please let me know your opinion on the border / outline colors, I'm not sure there's an example in our library that handles anything quite like this yet, happy to set it to whatever. Also the name of the component itself I went back and forth on; if you have a preference that's not what I have here I'm happy to change that as well.

## Testing

Checkout the storybook, validate that it works

## Issues

- Closes #255

### CODE REVIEW

**Comment by psirenny** [comment]
Will take a look ASAP.

File: packages/ui/src/components/one-time-passcode-input.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  HTMLAttributes<HTMLDivElement> & SlotProps
+>(({ char: character, className, isActive, ...properties }, reference) => {
+  const mergedClassName = cx(
+    "border-neutral-a-7 outline-primary-8 group-focus-within:border-neutral-a-8 group-hover:border-neutral-a-8 relative flex h-14 w-10 items-center justify-center border-y text-2xl outline outline-0 outline-offset-0 transition duration-300 ltr:border-r ltr:first:rounded-l-md ltr:first:border-l ltr:last:rounded-r-md rtl:border-l rtl:first:rounded-r-md rtl:first:border-r rtl:last:rounded-l-md",
+    isActive && "z-10 outline-2 ltr:border-l rtl:border-r",
+    className,
+  );
+  return (
+    <div className={mergedClassName} {...properties} ref={reference}>
+      {character !== null && <div>{character}</div>}
```
Comment (psirenny) [inline packages/ui/src/components/one-time-passcode-input.tsx:43]: Is the extra div required? I think I see it working without the wrapper.

File: packages/ui/src/components/one-time-passcode-input.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+);
+
+OneTimePasscodeInputDash.displayName = "OneTimePasscodeInputDash";
+
+export const OneTimePasscodeInputSlot = forwardRef<
+  HTMLDivElement,
+  HTMLAttributes<HTMLDivElement> & SlotProps
+>(({ char: character, className, isActive, ...properties }, reference) => {
+  const mergedClassName = cx(
+    "border-neutral-a-7 outline-primary-8 group-focus-within:border-neutral-a-8 group-hover:border-neutral-a-8 relative flex h-14 w-10 items-center justify-center border-y text-2xl outline outline-0 outline-offset-0 transition duration-300 ltr:border-r ltr:first:rounded-l-md ltr:first:border-l ltr:last:rounded-r-md rtl:border-l rtl:first:rounded-r-md rtl:first:border-r rtl:last:rounded-l-md",
```
Comment (psirenny) [inline packages/ui/src/components/one-time-passcode-input.tsx:37]: Could we give the slots a subtle background color in light and dark mode? Something like `bg-white-a-3 dark:bg-black-a-3`.

**Review by psirenny — CHANGES_REQUESTED** [review]
Nicely done, left some comments.

**Comment by psirenny** [comment]
@eakbulut @ThomasSeaver I went scouring Refero to see what others have done and happened upon these two examples which I thought were nice.

**Luma** (Surface input)
Screenshot 2024-05-31 at 9 15 34 AM

**Linktree* (Soft input)
Screenshot 2024-05-31 at 9 16 54 AM

Both of those have similar input variant styles as us: Soft and Surface. I also found Linktrees use of "0" as a placeholder very interesting. I don't know whether I love it or hate it 😛.

File: packages/ui/src/components/one-time-passcode-input.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (ThomasSeaver) [reply packages/ui/src/components/one-time-passcode-input.tsx:37]: I'm fine with whatever you think is best; I was just following the example. One issue with outlines is I don't believe they have a right / left / top / bottom width control, so you end up with a thick inner wall on the slots like this: 
image
which is something the example was avoiding by controlling the borders, though it was a little janky

I agree that it clashes, but I also think this looks a little funky:
image
Is this what you were imagining, or something different?

[the remark this answers is not in the document]
Comment (psirenny) [reply packages/ui/src/components/one-time-passcode-input.tsx:37]: Oh yea, that looks off. I think we'd need to make the the outline inset in that case. Can the offset be negative? `-outline-offset-px`

**Review by psirenny — APPROVED** [review]
Nice!

File: packages/ui/src/components/one-time-passcode-input.tsx
Comment (eakbulut) [inline packages/ui/src/components/one-time-passcode-input.tsx:1]: A couple of things:

- I think we need an error state when someone enters anything but a number, at the moment there is no feedback when the user does something wrong. Just turning the cell outline to red may suffice.

- Also related to the previous one; if I paste a 6 digit number it works fine. If I paste a 6 digit number with a hyphen in the middle, nothing happens. So, having a hyphen in the UI but not accepting an input with a hyphen is a bit conflicting. We should either remove the hyphen from the UI or disregard the hyphen and input rest of the code pasted without the hyphen or point out to the user that something is wrong.

- When I enter the code and want to go back to a specific digit to edit, it doesn't really work. When I click on first digit 3rd, 4rd or 6th digit gets the focus randomly.

- When I enter more than 6 digits, it keep updating the 6th digit with the extra numbers. e.g. If I accidentally enter `12345678`, I end up with `123458`. The focus should move away from 6th digit once it's filled.

**Review by eakbulut — CHANGES_REQUESTED** [review]

File: packages/ui/src/components/one-time-passcode-input.tsx
Comment (psirenny) [reply packages/ui/src/components/one-time-passcode-input.tsx:1]: 👀 🥹

**Comment by psirenny** [comment]
One thing I noticed also but I don't know if it requires addressing right away, is that the border between each slot should probably just be 1px instead of 2px. But I know that's complicated by the fact that we're using outlines with transparency.

**Comment by ThomasSeaver** [comment]
> I think we need an error state when someone enters anything but a number, at the moment there is no feedback when the user does something wrong. Just turning the cell outline to red may suffice.

Should do this now for all non-numerics, besides a hyphen which it should silently swallow

> Also related to the previous one; if I paste a 6 digit number it works fine. If I paste a 6 digit number with a hyphen in the middle, nothing happens. So, having a hyphen in the UI but not accepting an input with a hyphen is a bit conflicting. We should either remove the hyphen from the UI or disregard the hyphen and input rest of the code pasted without the hyphen or point out to the user that something is wrong.

This should work now by disregarding the hyphen when pasted / input, though I did have to bastardize the primitive a tiny bit. As a side note, l'm not against a different pattern for the input. At this point I want to get it in, but if we wanted to revisit this with a spec in hand and remove the hyphen / alter the slot pattern I'd be okay with that. 

> When I enter the code and want to go back to a specific digit to edit, it doesn't really work. When I click on first digit; 3rd, 4rd or 6th digit gets the focus randomly.

This one is a real limitation of the base component, since the way its patterned is that its an invisible fake input underneath. I think pursuing a change here would be really pushing the boundaries of 'using' the component and also a can of worms, so I didn't go after it. It MIGHT be possible, but because the input ref is created inside the primitive and it has its own selection state, it concerns me.

> When I enter more than 6 digits, it keeps updating the 6th digit with the extra numbers. e.g. If I accidentally enter 12345678, I end up with 123458. The focus should move away from 6th digit once it's filled.

It should now not re-update the sixth digit. As a secondary bit I added a 'confirm' button to the story to show how an `onComplete` handler could shift the focus with refs, since I don't think it's a responsibility of the base input component. I also envision other patterns, like auto submitting the code, etc

> One thing I noticed also but I don't know if it requires addressing right away, is that the border between each slot should probably just be 1px instead of 2px. But I know that's complicated by the fact that we're using outlines with transparency.

yeah, like I mentioned earlier because the outline doesn't have sided controls AFAIK having an outline on each slot means that there's a thick inner wall. If you have a solution to that I'm happy to implement it but I wasn't sure what other option I had if we wanted to use outlines

**Pull request merged**